MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1771 @@
+        }
+      }
+
+      sv_total_track_which_trkidx.push_back(track_idx);
+      sv_total_ntrack.push_back(ntracks);
+      sv_match_tracktojet_which_trkidx.push_back(tracktojet_which_trkidx);
+      sv_match_tracktojet_which_jetidx.push_back(tracktojet_which_jetidx);
+      vtxidx++;
+    }
+
+
+    if (vertices->size() >= 2) {
+
+      for (size_t i = 0; i < sv_ascending_vtxidx.size(); i++) {
+        const size_t vtxidx0 = sv_ascending_vtxidx[i];
+        reco::Vertex& sv0 = vertices->at(vtxidx0);
+        double sv0x = sv0.x() - bsx;
+        double sv0y = sv0.y() - bsy;
+        double phi0 = atan2(sv0y, sv0x);
+        if (verbose){
+          printf("-----loop # %lu ----- \n", i);
+          printf(" sv0'idx: %lu \n", vtxidx0);
+          printf(" sv0'ntrack: %u \n", sv0.nTracks());
+        }
+        for (size_t j = 0; j < sv_ascending_vtxidx.size(); j++) {
+          if (i == j) continue;
+          const size_t vtxidx1 = sv_ascending_vtxidx[j];
+          reco::Vertex& sv1 = vertices->at(vtxidx1);
+          if (verbose){
+            printf(" sv1'idx: %lu \n", vtxidx1);
+            printf(" sv1'ntrack: %u \n", sv1.nTracks());
+          }
+
+          // only consider a pair with at least 3 tracks per vertex
+          if (sv0.nTracks() > 2 && sv1.nTracks() > 2) {
+
+            std::pair<bool, std::vector<std::vector<size_t>>> sharedjet_tool = sharedjets(vtxidx0, vtxidx1, sv_match_tracktojet_which_jetidx, sv_match_tracktojet_which_trkidx);
+
+            // loop thru {1,1}+{1,n} nsharedjets and remove just one shared track from v0 if a |dPhi(v0,one shared track)| > pi/2
+            if (sharedjet_tool.first) {
+              if (verbose)
+                printf("start shj implementation to {1,1} and {1,n} \n");
+              std::vector<std::vector<size_t>> sv_lonesharedtrack_which_trkidx = sharedjet_tool.second;
+              std::vector<size_t> sv0_lonesharedtrack_which_trkidx = sv_lonesharedtrack_which_trkidx[0];
+              std::vector<size_t> sv1_lonesharedtrack_which_trkidx = sv_lonesharedtrack_which_trkidx[1];
+              if (verbose) {
+                printf("size of set of lone shared tracks per sv0: %lu \n", sv0_lonesharedtrack_which_trkidx.size());
+                printf("size of set of lone shared tracks per sv1: %lu \n", sv1_lonesharedtrack_which_trkidx.size());
+              }
+              for (size_t k = 0; k < sv0_lonesharedtrack_which_trkidx.size(); k++) {
+                track_vec tks_sv0 = sv_total_track_which_trk_vec[vtxidx0];
+                size_t idx = sv0_lonesharedtrack_which_trkidx[k];
+                h_resolve_shared_jets_lonetrkvtx_dphi->Fill(fabs(reco::deltaPhi(tks_sv0[idx]->phi(), phi0)));
+
+                // drop the lone track pointing backwards from the vertex direction!
+                if (fabs(reco::deltaPhi(tks_sv0[idx]->phi(), phi0)) > M_PI / 2) {
+                  eraseElement(sv_total_track_which_trkidx[vtxidx0], idx);
+                }
+              }

```

Comment:
What in all of this forces sv0 to be the one with a lone shared jet track, rather than sv1? And does `sv0_lonesharedtrack_which_trkidx.size()` have to be 1 for all of this to be valid? Maybe we explicitly should add an `if` statement corresponding to that. I worry that if we change any of the code that comes before this, it's possible that sv0 would no longer be forced to be the one with a lone shared jet track (and that sv1 could instead). We should definitely add comments about that, and protection in the code (this might be a good case to throw an error if our assumptions are violated).

Reply:
I think it is already expected that only sv0 is considered being dropped a lone shared track for each pair we consider their (multiple) shared jets. Since the sv1 from the earlier round can become the sv0 in the later round (based on the way we decided to write the double loops with double counting pairs). As long as the pair is no long sharing jet(s), the sv1 won't be removed 'lone shared-jet tracks' from any 'special' shared jets the pair has. The sv1 will be removed 'lone shared-jet tracks' if sv0 has all  'lone shared-jet tracks' being compatible to its vertex. 

So, to answer your question:
> does sv0_lonesharedtrack_which_trkidx.size() have to be 1 for all of this to be valid? 
the fact is that we can have multiple shared jets with {1,1} and {1,n} configs so this vector can have a size larger than one but all of tracks it contains are 'lone shared-jet tracks.'

Reply:
Clarify by writing a function for this part.